Encoder/DefaultEncoder.php

```diff
@@ -0,0 +1,163 @@
+     * @var string
+     */
+    protected $encryptionEngine;
+
+    /**
+     * @var string
+     */
+    protected $encryptionAlgorithm;
+
+    /**
+     * Constructor.
+     *
+     * @param KeyLoaderInterface $keyLoader
+     * @param string             $encryptionEngine
+     * @param string             $encryptionAlgorithm
+     *
+     * @throws InvalidArgumentException If the given encryptionEngine is not supported.
+     * @throws InvalidArgumentException If the given algorithm is not supported.
+     */
+    public function __construct(KeyLoaderInterface $keyLoader, $encryptionEngine, $encryptionAlgorithm)
+    {
+        if (!in_array($encryptionEngine, ['OpenSSL', 'SecLib'])) {
+            throw new InvalidArgumentException(
+                sprintf('The encryption engine "%s" is not supported', $encryptionEngine)
+            );
+        }
+
+        if (!$this->isAlgorithmSupportedForEngine($encryptionEngine, $encryptionAlgorithm)) {
+            throw new InvalidArgumentException(
+                sprintf('The algorithm "%s" is not supported for %s', $encryptionAlgorithm, $encryptionEngine)
+            );
+        }
+
+        $this->keyLoader           = $keyLoader;
+        $this->encryptionEngine    = $encryptionEngine;
+        $this->encryptionAlgorithm = $encryptionAlgorithm;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \RuntimeException If the JWS cannot be signed
+     */
+    public function encode(array $data)
+    {
+        $jws = new JWS(['alg' => $this->encryptionAlgorithm], $this->encryptionEngine);
+        $key = $this->keyLoader->loadKey('private');
+
+        $jws->setPayload($data);
+
+        try {
+            $jws->sign($key, $this->keyLoader->getPassphrase());
+        } catch (InvalidArgumentException $e) {
+            throw new \RuntimeException(
+                sprintf('An error occurred while signing the JWS (%s). Please check your configuration.', $e->getMessage())
+            );
+        }
+
+        return $jws->getTokenString();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws InvalidArgumentException If the JWS cannot be loaded
+     * @throws \RuntimeException        If the JWS cannot be verified
+     */
+    public function decode($token)
+    {
+        try {
+            $jws = JWS::load($token, false, null, $this->encryptionEngine);
+        } catch (InvalidArgumentException $e) {
+            return false;
+        }
+
+        $key     = $this->keyLoader->loadKey('public');
+        $payload = $jws->getPayload();
+
+        try {
+            if (!$jws->verify($key, $this->encryptionAlgorithm) || $this->isJWSExpired($payload)) {
+                return false;
+            }
+        } catch (InvalidArgumentException $e) {
+            throw new \RuntimeException(
+                sprintf('An error occurred while verifying the JWS (%s). Please check your configuration.', $e->getMessage())
+            );
+        }
+
+        return $payload;
+    }
```

Comment:
Agreed. The return of this method is handled in the [JWTProvider](https://github.com/lexik/LexikJWTAuthenticationBundle/blob/master/Security/Authentication/Provider/JWTProvider.php#L64), it throw the same AuthenticationException for the 3 possible cases you given ("Invalid token"). We can easily give a more precise message by creating these 3 exceptions and separate the different checks that would throw the corresponding exception.
But, as the JWTProvider needs to handle these exceptions properly (throw the same AuthenticationException but by catching the new exceptions and replace "invalid token" by the exception message for instance, instead of just wait for a bool), don't you think that this should be done in a next PR? Sort as this one doesn't touch the JWTProvider at all.